source/common/protobuf/deterministic_hash.cc

```diff
@@ -0,0 +1,228 @@
+      int c = reflection.FieldSize(message, &field);
+      for (int i = 0; i < c; i++) {
+        seed = HashUtil::xxHash64Value(reflection.GetRepeatedEnumValue(message, &field, i), seed);
+      }
+    } else {
+      seed = HashUtil::xxHash64Value(reflection.GetEnumValue(message, &field), seed);
+    }
+    break;
+  case FieldDescriptor::CPPTYPE_STRING: {
+    if (field.is_repeated()) {
+      for (const std::string& str : reflection.GetRepeatedFieldRef<std::string>(message, &field)) {
+        seed = HashUtil::xxHash64(str, seed);
+      }
+    } else {
+      std::string scratch;
+      seed = HashUtil::xxHash64(reflection.GetStringReference(message, &field, &scratch), seed);
+    }
+  } break;
+  case FieldDescriptor::CPPTYPE_MESSAGE:
+    if (field.is_map()) {
+      seed = reflectionHashMapField(message, field, seed);
+    } else if (field.is_repeated()) {
+      for (const Protobuf::Message& submsg :
+           reflection.GetRepeatedFieldRef<Protobuf::Message>(message, &field)) {
+        seed = reflectionHashMessage(submsg, seed);
+      }
+    } else {
+      seed = reflectionHashMessage(reflection.GetMessage(message, &field), seed);
+    }
+    break;
+  }
+  return seed;
+}
+
+// Converts from type urls OR descriptor full names to descriptor full names.
+// Type urls are as used in envoy yaml config, e.g.
+// "type.googleapis.com/envoy.extensions.filters.udp.udp_proxy.v3.UdpProxyConfig"
+// becomes
+// "envoy.extensions.filters.udp.udp_proxy.v3.UdpProxyConfig"
+absl::string_view typeUrlToDescriptorFullName(absl::string_view url) {
+  const size_t pos = url.rfind('/');
+  if (pos != absl::string_view::npos) {
+    return url.substr(pos + 1);
+  }
+  return url;
+}
+
+std::unique_ptr<Protobuf::Message> unpackAnyForReflection(const ProtobufWkt::Any& any) {
+  const Protobuf::Descriptor* descriptor =
+      Protobuf::DescriptorPool::generated_pool()->FindMessageTypeByName(
+          typeUrlToDescriptorFullName(any.type_url()));
+  // If the type name refers to an unknown type, we treat it the same as other
+  // unknown fields - not including its contents in the hash.
+  if (descriptor == nullptr) {
+    return nullptr;
+  }
+  const Protobuf::Message* prototype =
+      Protobuf::MessageFactory::generated_factory()->GetPrototype(descriptor);
+  std::unique_ptr<Protobuf::Message> msg(prototype->New());
+  any.UnpackTo(msg.get());
+  return msg;
+}
+
+// This is intentionally ignoring unknown fields.
+uint64_t reflectionHashMessage(const Protobuf::Message& message, uint64_t seed) {
+  using Protobuf::FieldDescriptor;
+  std::string scratch;
+  const Protobuf::Reflection* reflection = message.GetReflection();
+  const Protobuf::Descriptor* descriptor = message.GetDescriptor();
+  seed = HashUtil::xxHash64(descriptor->full_name(), seed);
+  if (descriptor->well_known_type() == Protobuf::Descriptor::WELLKNOWNTYPE_ANY) {
+    const ProtobufWkt::Any* any = Protobuf::DynamicCastToGenerated<ProtobufWkt::Any>(&message);
```

Comment:
Can this fail (i.e., return nullptr)? if so I think the code can return the seed.

Reply:
great catch Adi -- this is exactly why I wanted more eyes on this!

Adi can you suggest a test-case that might hit this? If we have to handle null here it'd be good to test it.

Reply:
Adding an ASSERT (it should not be able to fail because WELLKNOWNTYPE_ANY should always cast to Any.)